metricbeat/module/http/json/json.go

```diff
@@ -116,95 +115,50 @@ func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
 	}, nil
 }
 
-func (m *MetricSet) processBody(response *http.Response, jsonBody interface{}) common.MapStr {
-	var event common.MapStr
-
-	if m.deDotEnabled {
-		event = common.DeDotJSON(jsonBody).(common.MapStr)
-	} else {
-		event = jsonBody.(common.MapStr)
-	}
-
-	if m.requestEnabled {
-		event[mb.ModuleDataKey] = common.MapStr{
-			"request": common.MapStr{
-				"headers": m.getHeaders(response.Request.Header),
-				"method":  response.Request.Method,
-				"body": common.MapStr{
-					"content": m.body,
-				},
-			},
-		}
-	}
-
-	if m.responseEnabled {
-		phrase := strings.TrimPrefix(response.Status, strconv.Itoa(response.StatusCode)+" ")
-		event[mb.ModuleDataKey] = common.MapStr{
-			"response": common.MapStr{
-				"code":    response.StatusCode,
-				"phrase":  phrase,
-				"headers": m.getHeaders(response.Header),
-			},
-		}
-	}
-
-	// Set dynamic namespace
-	event["_namespace"] = m.namespace
-
-	return event
-}
-
-// Fetch methods implements the data gathering and data conversion to the right format
-// It returns the event which is then forward to the output. In case of an error, a
-// descriptive error must be returned.
-func (m *MetricSet) Fetch() ([]common.MapStr, error) {
+// Fetch methods implements the data gathering and data conversion to the right
+// format. It publishes the event which is then forwarded to the output. In case
+// of an error set the Error field of mb.Event or simply call report.Error().
+func (m *MetricSet) Fetch(reporter mb.ReporterV2) error {
 	response, err := m.http.FetchResponse()
 	if err != nil {
-		return nil, err
+		return err
 	}
-	defer response.Body.Close()
-
-	var jsonBody common.MapStr
-	var jsonBodyArr []common.MapStr
-	var events []common.MapStr
+	defer func() {
+		if err := response.Body.Close(); err != nil {
+			m.Logger().Debug("error closing http body")
+		}
+	}()
 
 	body, err := ioutil.ReadAll(response.Body)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
 	if m.jsonIsArray {
-		err = json.Unmarshal(body, &jsonBodyArr)
-		if err != nil {
-			return nil, err
+		var jsonBodyArr []common.MapStr
+		if err = json.Unmarshal(body, &jsonBodyArr); err != nil {
+			return err
 		}
 
 		for _, obj := range jsonBodyArr {
 			event := m.processBody(response, obj)
-			events = append(events, event)
+
+			if reported := reporter.Event(event); !reported {
+				return errors.New("error reporting event")
```

Comment:
This is related to the discussion we need to have in https://github.com/elastic/beats/issues/11666

In general I don't think we should report an error here as it normally means the metricset or beat was stopped on purpose (which is not an error).